src/attr.rs

```diff
@@ -0,0 +1,165 @@
+    False,
+    /// The attribute is set to a [valid UTF-8 string](prim@str).
+    String(&'string str),
+    /// The attribute is set to a non-UTF-8 string.
+    Bytes(&'string [u8]),
+    /// The attribute is not specified.
+    Unspecified,
+}
+
+impl<'string> AttrValue<'string> {
+    /// Returns the state of an attribute by inspecting its [value](crate::Repository::get_attr)
+    /// by a [string](prim@str).
+    ///
+    /// As [`str`](prim@str) is guaranteed to contain only valid UTF-8, this function never returns
+    /// [`AttrValue::Bytes`].
+    pub fn from_string(value: Option<&'string str>) -> Self {
+        match unsafe { raw::git_attr_value(value.map_or(ptr::null(), |v| v.as_ptr().cast())) } {
+            raw::GIT_ATTR_VALUE_TRUE => Self::True,
+            raw::GIT_ATTR_VALUE_FALSE => Self::False,
+            raw::GIT_ATTR_VALUE_STRING => Self::String(value.unwrap()),
+            raw::GIT_ATTR_VALUE_UNSPECIFIED => Self::Unspecified,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Returns the state of an attribute by inspecting its [value](crate::Repository::get_attr_bytes)
+    /// by a [byte](u8) [slice].
+    pub fn from_bytes(value: Option<&'string [u8]>) -> Self {
+        match unsafe { raw::git_attr_value(value.map_or(ptr::null(), |v| v.as_ptr().cast())) } {
+            raw::GIT_ATTR_VALUE_TRUE => Self::True,
+            raw::GIT_ATTR_VALUE_FALSE => Self::False,
+            raw::GIT_ATTR_VALUE_STRING => {
+                let value = value.unwrap();
+                if let Ok(string) = str::from_utf8(value) {
+                    Self::String(string)
+                } else {
+                    Self::Bytes(value)
+                }
+            }
+            raw::GIT_ATTR_VALUE_UNSPECIFIED => Self::Unspecified,
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Compare two [`AttrValue`]s.
+///
+/// Note that this implementation does not differentiate [`AttrValue::String`] and
+/// [`AttrValue::Bytes`].
+impl PartialEq for AttrValue<'_> {
+    fn eq(&self, other: &AttrValue<'_>) -> bool {
+        match (self, other) {
+            (Self::True, AttrValue::True)
+            | (Self::False, AttrValue::False)
+            | (Self::Unspecified, AttrValue::Unspecified) => true,
+            (AttrValue::String(string), AttrValue::Bytes(bytes))
+            | (Self::Bytes(bytes), AttrValue::String(string)) => string.as_bytes() == *bytes,
+            (Self::String(left), AttrValue::String(right)) => left == right,
+            (Self::Bytes(left), AttrValue::Bytes(right)) => left == right,
+            _ => false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::AttrValue;
+    use std::ffi::CStr;
+    use std::os::raw::c_char;
+
+    extern "C" {
+        // libgit2 defines them as mutable, so they are also declared mutable here.
+        // However, libgit2 never mutates them, thus it's always safe to access them in Rust.
+        static mut git_attr__true: *const c_char;
```

Comment:
I'm getting linker errors for these since 65c0dccdf6640d2a832395f0c5d2db763f36628d when trying to `cargo test`:

```
  = note: ld.lld: error: undefined symbol: git_attr__true
          >>> referenced by attr.rs:135 (src/attr.rs:135)
          >>>               /tmp/git2-rs/target/debug/deps/git2-33fc7bd91ae7f3bc.tgtslt60c5cmi6e.rcgu.o:(git2::attr::tests::attr_value_from_string::hba9c77ae386af5fe)
          >>> referenced by attr.rs:140 (src/attr.rs:140)
          >>>               /tmp/git2-rs/target/debug/deps/git2-33fc7bd91ae7f3bc.tgtslt60c5cmi6e.rcgu.o:(git2::attr::tests::attr_value_from_bytes::h12449f8f751186bc)

          ld.lld: error: undefined symbol: git_attr__false
          >>> referenced by attr.rs:0 (src/attr.rs:0)
          >>>               /tmp/git2-rs/target/debug/deps/git2-33fc7bd91ae7f3bc.tgtslt60c5cmi6e.rcgu.o:(git2::attr::tests::attr_value_from_string::hba9c77ae386af5fe)
          >>> referenced by attr.rs:0 (src/attr.rs:0)
          >>>               /tmp/git2-rs/target/debug/deps/git2-33fc7bd91ae7f3bc.tgtslt60c5cmi6e.rcgu.o:(git2::attr::tests::attr_value_from_bytes::h12449f8f751186bc)

          ld.lld: error: undefined symbol: git_attr__unset
          >>> referenced by attr.rs:0 (src/attr.rs:0)
          >>>               /tmp/git2-rs/target/debug/deps/git2-33fc7bd91ae7f3bc.tgtslt60c5cmi6e.rcgu.o:(git2::attr::tests::attr_value_from_string::hba9c77ae386af5fe)
          >>> referenced by attr.rs:0 (src/attr.rs:0)
          >>>               /tmp/git2-rs/target/debug/deps/git2-33fc7bd91ae7f3bc.tgtslt60c5cmi6e.rcgu.o:(git2::attr::tests::attr_value_from_bytes::h12449f8f751186bc)
          collect2: error: ld returned 1 exit status
```

What am I doing wrong?

Reply:
@kim I'm not sure... I am not able to reproduce with `RUSTFLAGS=-Clink-args=-fuse-ld=lld cargo test`.

Reply:
@kim Could you share what linker flags/`rustc` flags/Cargo profile settings you use and maybe a verbose build output? I'm sorry for the breakage and willing to fix the problem.